client/src/fluid/FluidTokenizer.res

```diff
@@ -699,82 +739,60 @@ let rec toTokens' = (~parentID=None, e: E.t, b: Builder.t): Builder.t => {
-    | TNewline(_) =>
-      row := row.contents + 1
-      col := 0
-    | _ => col := col.contents + length
-    }
-    pos := pos.contents + length
-    ti
-  })
-}
-
-let validateTokens = (tokens: list<fluidToken>): list<fluidToken> => {
-  List.forEach(tokens, ~f=t => {
-    asserT("invalid token", String.length(T.toText(t)) > 0, ~debug=t)
-    ()
-  })
-  tokens
-}
-
-// Remove artifacts of the token generation process
-let tidy = (tokens: list<fluidToken>): list<fluidToken> =>
-  tokens |> List.filter(~f=x =>
-    switch x {
-    | TIndent(0) => false
-    | _ => true
-    }
-  )
-
-let tokenizeWithFFTokenization = (
-  ffTokenization: featureFlagTokenization,
-  e: FluidExpression.t,
-): list<tokenInfo> =>
+let tokenizeExprWithOptions = (ffTokenization: featureFlagTokenization, e: E.t): list<tokenInfo> =>
   {...Builder.empty, ffTokenization: ffTokenization}
-  |> toTokens'(e)
+  |> exprToTokens(e)
   |> Builder.asTokens
   |> tidy
   |> validateTokens
   |> infoize
 
-let tokenize: E.t => list<FluidToken.tokenInfo> = tokenizeWithFFTokenization(
+@ocaml.doc("The 'default' tokenizer used in the main editor and basically
+  everywhere except for the feature flag editor")
+let tokenizeExpr: E.t => list<FluidToken.tokenInfo> = tokenizeExprWithOptions(
   FeatureFlagOnlyDisabled,
 )
 
-let tokensForEditor = (e: FluidTypes.Editor.t, ast: FluidAST.t): list<FluidToken.tokenInfo> =>
+@ocaml.doc("The 'default' tokenizer used in the main editor and basically
+  everywhere except for the feature flag editor")
+let tokenizeMatchPattern = (matchId: ID.t, index: int, mp: MP.t): list<tokenInfo> =>
+  matchPatternToTokens(matchId, ~idx=index, mp) |> tidy |> validateTokens |> infoize
+
+let tokenizeExprForEditor = (e: FluidTypes.Editor.t, expr: E.t): list<FluidToken.tokenInfo> =>
   switch e {
   | NoEditor => list{}
-  | MainEditor(_) => tokenize(FluidAST.toExpr(ast))
-  | FeatureFlagEditor(_, id) =>
-    FluidAST.findExpr(id, ast)
-    |> Option.map(~f=tokenizeWithFFTokenization(FeatureFlagConditionAndEnabled))
-    |> recoverOpt(
-      "could not find expression id = " ++ (ID.toString(id) ++ " when tokenizing FF editor"),
-      ~default=list{},
-    )
+  | MainEditor(_) => tokenizeExpr(expr)
+  | FeatureFlagEditor(_) => tokenizeExprWithOptions(FeatureFlagConditionAndEnabled, expr)
+  }
+
+let tokenizeMatchPatternForEditor = (
+  e: FluidTypes.Editor.t,
+  matchID: ID.t,
+  index: int,
+  mp: MP.t,
+): list<FluidToken.tokenInfo> =>
+  switch e {
+  | NoEditor => list{}
+  | MainEditor(_) => tokenizeMatchPattern(matchID, index, mp)
+  | FeatureFlagEditor(_) => tokenizeMatchPattern(matchID, index, mp)
   }
 
-let tokenizeForEditor = (e: FluidTypes.Editor.t, expr: FluidExpression.t): list<
+// this is only used for FluidDebugger. TODO: consider removing? Not sure why
+// we have this separate tokenizer - when would we have a FF but no expr
+// corresponding to the expr's ID?
+let tokenizeExprForDebugger = (e: FluidTypes.Editor.t, ast: FluidAST.t): list<
```

Comment:
outstanding question - why do we need tokenizer?
when would we have a FF but no expr corresponding to the expr's ID?